src/main/java/org/opentripplanner/updater/vehicle_positions/VehiclePositionPatternMatcher.java

```diff
@@ -0,0 +1,142 @@
+package org.opentripplanner.updater.vehicle_positions;
+
+import com.google.transit.realtime.GtfsRealtime.VehiclePosition;
+import org.opentripplanner.index.model.RealtimeVehiclePosition;
+import org.opentripplanner.model.FeedScopedId;
+import org.opentripplanner.model.Trip;
+import org.opentripplanner.routing.edgetype.TripPattern;
+import org.opentripplanner.routing.graph.Graph;
+import org.opentripplanner.routing.graph.GraphIndex;
+import org.opentripplanner.routing.vertextype.TransitStop;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Responsible for converting vehicle positions in memory to exportable ones, and associating each position
+ * with a pattern.
+ */
+public class VehiclePositionPatternMatcher {
+    private static final Logger LOG =
+            LoggerFactory.getLogger(VehiclePositionPatternMatcher.class);
+
+    private final GraphIndex graphIndex;
+
+    public VehiclePositionPatternMatcher(Graph graph) {
+        graphIndex = graph.index;
+    }
+
+    /**
+     * Iterates over a pattern's vehicle positions and removes all not in a new list of vehicle positions
+     * @param pattern           Pattern to "clean"
+     * @param vehiclePositions  List of new vehicle positions
+     */
+    public void cleanPatternVehiclePositions(TripPattern pattern, List<VehiclePosition> vehiclePositions) {
+        for (String key : pattern.vehiclePositions.keySet()) {
+            for (VehiclePosition vehiclePosition : vehiclePositions) {
+                String tripId = vehiclePosition.getTrip().getTripId();
+                if (tripId.equals(key)) {
+                    break;
+                }
+            }
+            // If we didn't return, it means this vehicle has ceased to exist and should be removed
+            pattern.vehiclePositions.remove(key);
+        }
+    }
```

Comment:
This is somewhat inefficient since it is O(n * m). However, a more concerning thing is that all vehicle positions will always be removed since the `break;` statement on line 43 only breaks out of the loop beginning on line 40 instead of the line 39 loop.

A more efficient way to do this would be to create a set of seen trip Ids while applying updates. After those seen trip IDs have been calculated, then pattern vehicle positions that weren't in the seen trip IDs could be removed.  If using a HashSet this would make the computational complexity O(n). Example:

```suggestion
    public void cleanPatternVehiclePositions(TripPattern pattern, Set<String> seenTripIds) {
        for (String key : pattern.vehiclePositions.keySet()) {
            if (!seenTripIds.contains(key)) {
                pattern.vehiclePositions.remove(key);
            }
        }
    }
```

Reply:
Yeah I was worried about this yesterday as I was committing. This is exactly the clever solution I was looking for thank you! Implemented in 41184c91d3a95efcf908bcc5c05ef1cb3b465043